src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,416 @@
+    }
+  }
+  // write last aggregate group
+  _set_and_write_aggregate_value<ColumnType, AggregateType, function>(
+      aggregate_results, aggregate_null_values, aggregate_group_index, aggregate_index, current_aggregate_value,
+      value_count, value_count_with_null, unique_values);
+
+  _output_segments[aggregate_index + _groupby_column_ids.size()] =
+      std::make_shared<ValueSegment<AggregateType>>(aggregate_results, aggregate_null_values);
+}
+
+template <typename ColumnType, typename AggregateType, AggregateFunction function>
+void AggregateSort::_set_and_write_aggregate_value(
+    std::vector<AggregateType>& aggregate_results, std::vector<bool>& aggregate_null_values,
+    uint64_t aggregate_group_index, [[maybe_unused]] uint64_t aggregate_index,
+    std::optional<AggregateType>& current_aggregate_value, [[maybe_unused]] uint64_t value_count,
+    [[maybe_unused]] uint64_t value_count_with_null, const std::unordered_set<ColumnType>& unique_values) const {
+  if constexpr (function == AggregateFunction::Count) {
+    if (this->_aggregates[aggregate_index].column) {
+      current_aggregate_value = value_count;
+    } else {
+      current_aggregate_value = value_count_with_null;
+    }
+  }
+  if constexpr (function == AggregateFunction::Avg &&
+                std::is_arithmetic_v<AggregateType>) {
+    if (value_count == 0) {
+      current_aggregate_value = std::optional<AggregateType>();
+    } else {
+      current_aggregate_value = *current_aggregate_value / value_count;
+    }
+  }
+  if constexpr (function == AggregateFunction::CountDistinct) {
+    current_aggregate_value = unique_values.size();
+  }
+  aggregate_null_values[aggregate_group_index] = !(current_aggregate_value.has_value());
+  if (current_aggregate_value.has_value()) {
+    aggregate_results[aggregate_group_index] = *current_aggregate_value;
+  }
+}
+
+std::shared_ptr<const Table> AggregateSort::_on_execute() {
+  const auto input_table = input_table_left();
+
+  // check for invalid aggregates
+  for (const auto& aggregate : _aggregates) {
+    if (!aggregate.column) {
+      if (aggregate.function != AggregateFunction::Count) {
+        Fail("Aggregate: Asterisk is only valid with COUNT");
+      }
+    } else {
+      DebugAssert(*aggregate.column < input_table->column_count(), "Aggregate column index out of bounds");
+      if (input_table->column_data_type(*aggregate.column) == DataType::String &&
+          (aggregate.function == AggregateFunction::Sum || aggregate.function == AggregateFunction::Avg)) {
+        Fail("Aggregate: Cannot calculate SUM or AVG on string column");
+      }
+    }
+  }
+
+  // collect groupy column definitions
+  for (const auto &column_id : _groupby_column_ids) {
+    _output_column_definitions.emplace_back(input_table->column_name(column_id),
+                                            input_table->column_data_type(column_id), true);
+  }
+
+  // collect aggregate column definitions
+  ColumnID column_index{0};
```

Comment:
We believe it cannot be scoped, because then we would create a new variable each time.
We prefer the current loop over the indexed for loop for readability reasons (and we saw both versions in other hyrise code).